pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go

```diff
@@ -0,0 +1,235 @@
+func (c Controller) getActiveNodeSets(nodes []int) map[string]sets.Int {
+	nodeSets := make(map[string]sets.Int)
+	states := c.numaStater.GetMachineState()
+	for _, node := range nodes {
+		nodeState := states[node]
+		for podUID := range nodeState.PodEntries {
+			if _, ok := nodeSets[podUID]; !ok {
+				nodeSets[podUID] = sets.Int{}
+			}
+			nodeSets[podUID].Insert(node)
+		}
+	}
+	return nodeSets
+}
+
+// todo: to leverage the active group MB of whole package to save metric store fetching
+func (c Controller) getNodeMBMetrics(nodes sets.Int) (strategy.GroupMB, error) {
+	mbs := make(map[int]float64)
+	for node := range nodes {
+		nodeMB, err := c.metricReader.GetNumaMetric(node, "mb-mbps")
+		if err != nil {
+			return nil, err
+		}
+		mbs[node] = nodeMB.Value
+	}
+	return mbs, nil
+}
+
+func (c Controller) getActiveGroupMBs(nodes []int) (toSkip bool, mbs strategy.GroupMBs, err error) {
+	activeGroups := c.getActiveNodeSets(nodes)
+	if len(activeGroups) <= 1 {
+		// no need to take any action; neither an error
+		toSkip = true
+		return
+	}
+
+	mbs, err = c.getGroupMBMetrics(activeGroups)
+	return
+}
+
+func (c Controller) getGroupMBMetrics(actives map[string]sets.Int) (strategy.GroupMBs, error) {
+	groups := make(strategy.GroupMBs, 0)
+	for _, nodes := range actives {
+		groupMB, err := c.getNodeMBMetrics(nodes)
+		if err != nil {
+			return nil, err
+		}
+		groups = append(groups, groupMB)
+	}
+	return groups, nil
+}
+
+func (c Controller) processPackage(packageID int, nodes []int) {
+	// get the metrics
+	currMetric, err := c.metricReader.GetPackageMetric(packageID, "rw-mbps")
+	if err != nil {
+		c.metricEmitter.StoreInt64("package metric reading error", 1, metrics.MetricTypeNameCount)
+		return
+	}
+
+	mbwPackage := int64(currMetric.Value)
+
+	// workloads are hi-prio (lo-prio not in scope of this stage)
+	// adjust mem bandwidth based on mbw metrics, if applicable
+	if mbwPackage >= c.bandwidthThreshold/100*MEMORY_BANDWIDTH_PHYSICAL_NUMA_PAINPOINT {
+		// over the threshold - to throttle noisy neighbors
+		// identify the numa-node-set (nodes of one workload) of this package
+		toSkip, activeGroupMBs, err := c.getActiveGroupMBs(nodes)
+		if err != nil {
+			// error happened; skip for now
+			c.metricEmitter.StoreInt64("numa metric reading error", 1, metrics.MetricTypeNameCount)
+			return
+		}
+
+		if toSkip {
+			return
+		}
+
+		shareAllocator := strategy.NewAllocator()
+		shares := shareAllocator.AllocateDeductions(activeGroupMBs, float64(mbwPackage), float64(c.bandwidthThreshold))
+		for _, group := range shares {
+			for node, share := range group {
+				// ignore if the deduction step is too small
+				if share >= c.minDeductionStep {
+					c.numaThrottled.Insert(node)
+					if err := c.mbAdjust.AdjustNumaMB(node,
+						uint64(activeGroupMBs.NodeMBAverage()),
+						uint64(share),
+						mbm.MEMORY_BANDWIDTH_CONTROL_REDUCE,
+					); err == nil {
+						c.numaThrottled.Insert(node)
```

Comment:
c.numaThrottled.Insert(node) has beed called at line 162? Is it duplicated?

Reply:
yes it is. line 168 should be kept as it has error checked. will update code.